src/test/run-make-fulldeps/coverage-reports-base/expected_show_coverage.async.txt

```diff
@@ -0,0 +1,66 @@
+    1|       |#![allow(unused_assignments)]
+    2|       |
+    3|       |// require-rust-edition-2018
+    4|       |
+    5|      1|async fn f() -> u8 { 1 }
+    6|       |
+    7|       |async fn foo() -> [bool; 10] { [false; 10] }
+    8|       |
+    9|       |pub async fn g(x: u8) {
+   10|       |    match x {
+   11|       |        y if f().await == y => (),
+   12|       |        _ => (),
+   13|       |    }
+   14|      1|}
```

Comment:
I think I'd like to count the opening brace or signature, as you suggest. I'll look into that.

There's a second issue here:

Why do the generator bodies show no counts? (Not even 0.)

I need to understand that and see if it can be fixed.

If I counted the bodies, then we SHOULD see something like:

```
    5|      1|async fn f() -> u8 { 1 }
                                                  ^0
```

As it is, though, there's no way to tell that this async function was actually NOT executed.

Reply:
Interesting. I wonder if the counters are getting optimized out somehow.

Reply:
Partial correction. The body of `f()` actually was executed (because the test blocks on `i()`, and `i()` `await`s `f()`.

But `g()` is a similar example of the point I was trying to make.

From the debug logs, I can see the generator/closure MIR for the body of `g()` is never passed to Rust's codegen phase. 

(FYI, -Clink-dead-code does not make a difference either; I think that flag is only used by the target linker.)

I think I may have looked into this (and if so, it was a hard problem) but it would be nice to find any MIR's skipped at codegen, and still inject something in the coverage map for the function body to, hopefully, produce line counts of 0 for the body.

Reply:
@tmandry @wesleywiser - I'm a little unsure how to proceed here.

I'll try my best to summarize my understanding and what I think I want to do but don't know how:

1. Given a `Crate`, `rustc_mir::monomorphize::collector` finds the "root" `DefId`s (via the `RootCollector`, a HIR `ItemLikeVisitor`), and traverses them to generate a set of `MonoItem`s. This is where the async closures (`Generator`s) are getting dropped if they are not going to be called. (From debug logs, `i::{closure#0}()` and `f::{closure#0}()` are both logged, because they are going to be called, but the other async functions' closures are not mentioned at all. There is no "path" from the roots (only `main()` in this case) to those closures.

2. It seems to me I can make another `ItemLikeVisitor`, find all of the `hir::ItemKindFn`s, get their `DefId`s, and look them up in the set of `MonoItem`s. If a `DefId` is not found among the `MonoItem`s, then it will not be codegen'ed, and therefore won't have coverage.

3. At that point, I want to add at least one `Zero` counter with span (code region) for the span of the closure/generator, for each missing `DefId`, to the `rustc_codegen_ssa::coverageinfo::map::FunctionCoverage` (the collection of counters and coderegions built up during the codegen phase, and eventually written out as the "Coverage Map". I would do this with `FunctionCoverage::add_counter(self, counter, region)`.

Here's where I get confused...

The `FunctionCoverage` is stored in a hash map in `CrateCoverageContext`, and that object is constructed as a field in `CodegenCx`, but `CodegenCx` appears to be constructed for a *specific `CodegenUnit`*

IIUC, codegen of a given crate can be partitioned across multiple `CodegenUnit`s, and each `CodegenUnit` will have only a subset of the crate's `MonoItem`s; so if that's true:

1. Given a `DefId` in some `Crate`, even if the `DefId` is not be found among the `MonoItem`s of one `CodegenUnit`, the `DefId` might still be found among the `MonoItem`s of one of the other`CodegenUnit`s for that `Crate`. If so, that prevents me being able to check for missing `DefIds` when I'm starting to codegen the Coverage Map, because the Coverage Map finalization happens within the `CodegenCx` of only one `CodegenUnit`.

2. But if I search for a `DefId` among all `MonoItem`s of all `CodegenUnit`s, and don't find it, how do I know which `CodegenCx` to use, to add the missing Counter and code region for that function?

What's even more baffling to me is, it looks to me like a CoverageMap is codegenned for each `CodegenCx`, and assuming there are more than one, then the code appears to be set up to generate multiple coverage maps. I don't really get how that would work (or why it appears to work now), but I assume it does work since evidence suggests as much.

So I guess I just don't understand something, and hopefully the solution (to adding counters and code regions for functions/closures that are missing from the MonoItems is easier than it appears.

Reply:
The discussion continued at https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/counters.20for.20uncalled.20functions

Reply:
I couldn't simply inject code regions for functions not codegenned, as discussed in the PR for my attempted changes in draft PR #79392, but I was able to make other changes to improve this.

I'm now adding a coverage code region for the function signature, up to (but not including) the opening brace. The existing code for combining spans helps simplify (I think in a good way) how the function signature span can continue into the function body, if no other spans interfere. I think the results look good.

I also addressed showing `0` coverage for closures and `async` function bodies by injecting what the Coverage Mapping Format calls a "gap region". There are additional notes on this, and some non-intuitive results (but not wrong) occasionally. Given the trade-offs, this is better for ensuring missing coverage is identified.

And, as I showed in my earlier comment tonight, `-C link-dead-code` *does* make a difference for some unused functions. (Just pointing that out, since I discussed it here as well.)

Reply:
I made a different change, backing out my original approach to closures, and I'm now adding coverage for any unused MIR. This addresses missing coverage for async functions, closures, and unused regular functions.